### DISCUSSION TITLE
live.ppooll dev

### DISCUSSION THREAD

**Discussion by the-steech** [ppooll-dev/ppooll#81] [Ideas]
I've just created a new branch for live.ppooll-dev.  still probably very buggy, but i wanted to give @klausfilip and @chausch a chance to try and see my changes.

until it's a bit more stable, might be good to keep this separate from main

**Comment by klausfilip** [comment]
how to test this???
only with live or also backward compatibility?
should the branch work as live-less release as well?

**Comment by klausfilip** [comment]
your solution with mc-outs as subpatch looks good !!
maybe thats even more convenient for ppooll_in_max anyways.

**Comment by klausfilip** [comment]
i just downloaded a live trial version.
can you give a brief introduction how to run this with ppooll4live ?

**Reply by the-steech** [reply]
Yes!  You should be able to just drag the .amxd file onto a Live MIDI track.  You can route Audio in from other Live tracks.  Use INmulti to receive and route that incoming audio

**Comment by klausfilip** [comment]
amazing !!
so whats the points?
- ll.live.check_envi_type.js checks if its in live and opens acts as bpatchers.
- ll.movewindow is replaced by 11.olsens "move bpatcher" then, right?
- m4l somehow refuses to open the abstraction def_outputs, so we need to nest it in the patcher.
- audio between max and live are maintained by ho_st and INmulti only.
- some more ??
is there anything, that should not be there in the max-without-live version ?

**Reply by the-steech** [reply]
Thanks klaus!!!  Yes this is all correct.  

There are 2 main JavaScript files: ll.live.check_envi_type and ll.live.init_bpatcher.

Check_envi_type happens in an act that loads another act (right now, only ho_st and ll.buffermenu)

Init_bpatcher lives in the "actmaker" abstraction and sets the size of the acts bpatcher.

You're right about audio in via INmulti and audio out via ho_st.  The def outputs should appear nested in the INmulti if you hit the open button (may have to click twice, I think it's a bug with my implementation)

The last thing that needs is to somehow load/save envi files with the Ableton Live set.  Other than that, it's pretty much done I think!

**Comment by klausfilip** [comment]
ok !!
just tried to load a simple environment in live and it worked. (with the "old" system though)
anyway, we should try to merge our 2 branches ! (the hassle is ho_st, needs to be done manually i guess) 
and then continue from there.
what do you think?

**Reply by the-steech** [reply]
Maybe best move is to update the live.ppooll branch from main so that I can add your envi changes to live.ppooll version.

Then I can add back the live.ppooll changes to ho_st, etc, then test again.

How does that sound?

**Comment by klausfilip** [comment]
so you say, i should merge mine to main first?
i would prefer to this tested a bit before.

**Reply by the-steech** [reply]
ohhh, no i see!  in that case, should I? :
- merge my live.ppooll branch into your environment 
- add any missing changes from live.ppooll

and then we can test your newly merged environment changes branch from there?

**Comment by klausfilip** [comment]
yes, i would prefer this.
then we have a branch with all recent updates to test for us...

**Reply by the-steech** [reply]
if I have your ok @klausfilip, I will make the changes now!  just let me know :)

**Comment by klausfilip** [comment]
please !!
i just backuped my stuff

**Reply by the-steech** [reply]
@klausfilip all done-- everything seems to be working from my brief testing!!  i will be able to test further tomorrow :)

**Comment by klausfilip** [comment]
fantastic, 
first encounter:
- tetris does not change the window size.
- screen ON in an envi should be ignored (or always)

**Comment by the-steech** [comment]
Not at my computer atm, but there are a couple tweaks that need to be made (making list here to remind myself)

- hide host controls: audio on/off, screen, toolbar, # of output channels, etc
- set background color according to ppooll prefs

**Comment by klausfilip** [comment]
i worked on ll.mc.def_outputs::
to keep things sorted, there is a new file ll.mc.def_outputsBP
that provides it as bpatcher.
have a look into INmulti ... the def-button is now a toggle to open/close the bpatcher/size window.
can you live with this?

**Reply by the-steech** [reply]
yes this is working great!!

**Comment by the-steech** [comment]
@klausfilip when ho_st loads in live.ppooll, i added some "script hide" instructions to the init_bpatcher.js file.  here's what it looks like
<img width="159" alt="image" src="https://user-images.githubusercontent.com/114934553/218513800-0ba991f7-fa21-46cf-ad33-10ad4f9a9cbb.png">

**Reply by the-steech** [reply]
the only thing i'm not sure what to do about is the # of output channels.  live.ppooll only supports 16 ins/outs right now.  

should we automatically set the "chans" param to 16 when loading in live.ppooll?  would this have implications for someone trying to load an environment which was saved with more/less than 16 chans?

**Reply by klausfilip** [reply]
i think its ok to set 16 as default. if somebody sets to less than 16 there should not be a problem. if more i don't know, but i guess it would be just ignored...

**Reply by klausfilip** [reply]
wher is this init_bpatcher.js sitting?

**Reply by the-steech** [reply]
ahhh the proper name is "ll.live.initialize_bpatcher.js", in the /javascripts folder

**Reply by klausfilip** [reply]
i meant, where in the patches is it called/used?

**Reply by the-steech** [reply]
that .js is called in actmaker abstraction

**Comment by the-steech** [comment]
made a couple more tweaks.  simplified some of the code, removed brackets from param names (this causes issues with dict/json environments), and you can set patcher background from ppooll preferences sub!

**Reply by klausfilip** [reply]
which params, which brackets? some of them are essential !! like inputs for example..
also param-names of yourse are used in presets and envis.

**Reply by the-steech** [reply]
i think that conflicts can come up if the actual param name is like"gain[1]", because dicts/json use brackets to find the index of an element in a list.  for the live.params_in act, i changed all the params from "param[#]" to "param_#" to avoid this issue.  but that was the only change related to this!

**Reply by klausfilip** [reply]
well, most params[#] are leftovers, created by scripts etc. thats why i ask where you found them.
most likely the pattrs need to be deleted there anyways.

**Reply by klausfilip** [reply]
[#] also are created if you duplicate a GUI, and then they need a new name.

**Reply by the-steech** [reply]
yes that makes sense.  i think i ran into this issue when i was working on matriarch@.  i asked in the max discord, and they advised to avoid using "[#]" in for anything that may appear in a dict/json

**Comment by klausfilip** [comment]
@chausch the branch "environment speedup" is now merged with the newest hot suff from joe and me.
please give us a go to merge this into main ;)

**Reply by chausch** [reply]
hi guys, please excuse the silence, i'm super busy atm but will be "back" soon. 

just tested the enviroment rewrite. really speedy, great stuff. i could not find any errors from my half hour of testing, but have a small thought regarding the envi-load window being merged into console (which is a good thing imho):
@klausfilip  could there be some indicator as to the loading progress? e.g. a counter e.g. 3/12 in front of the act loading message when act 3 from 12 total acts in the enviroment was loaded? or maybe the loading initialization & finalization being split into 2 messages?

reason for this is that i have a ton of old enviroments that do not open correctly anymore (like, really really old ones). before there was the stuck envi load window reminding me of the fact that a certain act was tried to get loaded but couldn't. this way i could identify the culprit, check out the missing (or failing act) and re-import the enviroment. 
now there's just the confirmation after the load!

**Reply by klausfilip** [reply]
unclear to me:
the message
--------environment_loading-----: act modul.ator
is written before loading this act.
if it fails the loading stops there. (and you know there is a problem with the printed act)

if the act loads (responding with its "acting" message) the next act is loaded.

we could add a message after the act loaded - sth like 
----done----modul.ator1----
but that would be at the same moment as loading the next act.

**Comment by the-steech** [comment]
organizational question: do we want to move the .amxd to a separate repo under ppooll organization?  might help to keep development of ppooll and the .amxd patch separate

also, i'll start working on install/usage instructions

**Reply by klausfilip** [reply]
i am not sure. the cute thing is, that its only one .amxd !
and the rest is the same for both - no?
there is no better place than where it is now for my taste.

**Comment by chausch** [comment]
ps: i don't have ableton live, so can't test here, sorry, but great that this ppooll port for live exists, @the-steech! 🥳

**Reply by the-steech** [reply]
thanks so much @chausch :)  you actually don't need Live to try it, hypothetically you could open the live.ppooll.amxd file in regular Max and still get audio in/out.  the great thing about the port is that its basically just a bpatcher version of regular ppooll

**Reply by klausfilip** [reply]
..this solution is really fun.
before that i thought, we would need to make an amxd out of every act...

//
when i tested in live, the obvious thought came up, if i can have more live-tracks with different ppoolls ;)
which is currently not working resulting in a serious mess ;))))

**Reply by the-steech** [reply]
lololol 😂 i dont think it's ready for that yet :)  

curious question: is it possible to run 2 instances of ppooll in regular Max?  (ho_st1, ho_st2 ?).  i know klaus you mentioned that you can only run one buffer_host at a time

**Reply by klausfilip** [reply]
well, there is many global send/receives - assuming there will be always only one ho_st and one buffer_host.
would be another 3 weeks of sorting this out.

**Reply by the-steech** [reply]
maybe we want to consider some sort of prevention of this?  like, if ppooll_host/buffer_host is already open, don't allow Max to open another instance

**Reply by klausfilip** [reply]
sure -:|

**Comment by klausfilip** [comment]
ahm,
i actually have troubles with the new envi loading !!
getting crashes at parameter loading, which is hard to track.
but glad, you guys did not have this...

**Comment by klausfilip** [comment]
@the-steech  there is control@ with a live.ppooll mode

**Comment by klausfilip** [comment]
@the-steech i looked into this act-window issue, and added a workaround lieke this:
#############
 if (dict_act_sizes.contains(name)){
	var arr = dict_act_sizes.get(name)
	}
  else { //get the patching rect from file.
	var file_as_dict = new Dict("fad");
	file_as_dict.readany(prev.patcher.filepath);
	if (file_as_dict.contains("patcher::rect")){
		//post("soso");
		var arr = file_as_dict.get("patcher::rect");
		arr = [arr[0], arr[1], arr[2]+arr[0], arr[3]+arr[1]] //tja, the other rect-format ;)
		dict_act_sizes.set(name, arr);
		dict_act_sizes.export_json(HARDCODED_PATH + "act_sizes.json");	
	}
	}
######################

this should also solve the same issue, if an act is inside these other folders.

also i was looking for getting this patching_rect from the maxpat file without loading the eventually very BIG patcher file.
but did not find a way...

**Comment by klausfilip** [comment]
@the-steech 

[Archive.zip](https://github.com/ppooll-dev/ppooll/files/10798565/Archive.zip)
actually this would do it.
avoiding missing factory prefs, 
and also avoiding updating act_sizes when an act changes.

**Comment by klausfilip** [comment]
ll.live.initialize_bpatcher.js
?##########################

/*
  ll.live.initialize_bpatcher.js
  fka js.initializeRect.js
  by steech

    2nd step in a 2-step, act-loading process

    arguments:
      instance: (int) instance number, ie '1' in 'sinus1'
      name: (symbol) act name, ie 'sinus' in 'sinus1'

    1) user selects an act from ho_st
        ho_st sends the actname to the [environment] subpatch
        a bpatcher is created in the [environment] subpatch with the act .maxpat file

    2) this script runs within "actmaker" object in the newly loaded bpatcher
        find this bpatcher as an object within [environment] patch
        set bpatcher object size based on act tetris setting
        set 'movewind' jsui object at top-level of act to draggable jsui
        
        if this is the ho_st bpatcher, it will load additional acts 
          - live.midi_in
          - live.params_in
*/


autowatch = 1;

var HARDCODED_PATH = "~/Documents/Max 8/Packages/ppooll/misc/ppooll_factory_presets/"

var TO_HIDE = ['audioON/OFF','menubar','screen']

var IGNORE_ACTS_LIST = [
  'live.midi_in',
  'live.params_in',
]

var console = {
  log: function(message){
    post("ll.live.initialize_bpatcher: " + message)
    post()
  }
}

function make(name, instance) {
  // var name = arguments[0]            // ie 'sinus'
  // var instance = arguments[1]        // ie '1'

  var nameInstance = name + instance // ie 'sinus1'

  // console.log("create act " + nameInstance)

  var patcher = this.patcher.box;
  var prev = 0
  var owner = this.patcher.box

  // ignore acts that are meant to be hidden and will always load in environment
  if(IGNORE_ACTS_LIST.indexOf(name) > -1){
    // console.log('skip renaming "'+ nameInstance +'"')
    return
  }

  var isMaxForLive = false

  // iterate over parent patches to find the act-level bpatcher
  while ( owner ) {
    prev = owner
    owner = owner.patcher.box

    if(prev.patcher.name === name || prev.patcher.name === "ppooll_host")
      isMaxForLive = true
      break;
  }

  if(!isMaxForLive || !prev.patcher.box){
    // console.log('is Max runtime; set ll.movewindow.js')
    // reset movewindow ppooll jsui to ppooll default file (for dragging around desktop)
    if(prev.patcher.getnamed('master'))
      prev.patcher.message("script","sendbox","master", "filename", "ll.movewindow.js")

    if(prev.patcher.getnamed('movewind'))
      prev.patcher.message("script","sendbox","movewind", "filename", "ll.movewindow.js")

    return
  }

  // found the act's patcher box-- time to initialize...
  console.log('will rename bpatcher to "'+ nameInstance +'"')
  
  //set box varname to nameInstance
  prev.patcher.box.varname = nameInstance
  

  var coords = [0, 0, 200,200];
  coords = getcoords(prev.patcher.filepath);

/*
  var dict_act_sizes = new Dict("act_sizes")
  dict_act_sizes.import_json(HARDCODED_PATH + "act_sizes.json")
  if (dict_act_sizes.contains(name)){
	var arr = dict_act_sizes.get(name)
	}
  else { //get the patching rect from file.
	var file_as_dict = new Dict("fad");
	file_as_dict.readany(prev.patcher.filepath);
	if (file_as_dict.contains("patcher::rect")){
		//post("soso");
		var arr = file_as_dict.get("patcher::rect");
		arr = [arr[0], arr[1], arr[2]+arr[0], arr[3]+arr[1]] //tja, the other rect-format ;)
		dict_act_sizes.set(name, arr);
		dict_act_sizes.export_json(HARDCODED_PATH + "act_sizes.json");
		
	}
	}
  
  if(arr){
    var coords = [arr[0], arr[1], arr[2]-arr[0], arr[3]-arr[1]]
*/
    // set patching rect of act's bpatcher & bring to front
    owner.patcher.message("script","sendbox",nameInstance,"patching_rect",coords)
    owner.patcher.message("script","bringtofront",nameInstance)
  //}


  // locate draggable jsui corner for acts and replace with moveable_ui for bpatcher
  if(prev.patcher.getnamed('master'))
    prev.patcher.message("script","sendbox","master", "filename", "jsui_drag_bpatcher.js")

  if(prev.patcher.getnamed('movewind'))
    prev.patcher.message("script","sendbox","movewind", "filename", "jsui_drag_bpatcher.js")

  // initialize other acts if this is the ho_st
  if(name === 'ho_st'){
    console.log('create ho_st1')

    for(var i=0; i<TO_HIDE.length; i++){
      if(prev.patcher.getnamed(TO_HIDE[i])){
        prev.patcher.message("script","hide",TO_HIDE[i])
      }
    }

    // clear out old, live.ppooll-specific acts
    if( owner.patcher.getnamed("live.midi_in1") ){
      // console.log('delete live.midi_in')
      owner.patcher.message("script", "delete", "live.midi_in1")
    }
    if( owner.patcher.getnamed("live.params_in1") ){
      // console.log('delete live.params_in')
      owner.patcher.message("script", "delete", "live.params_in1")
    }

    // create live.ppooll-specific acts
    owner.patcher.message("script", "hidden", "newdefault", "live.midi_in1", 5, 91, "live.midi_in")
    owner.patcher.message("script", "hidden", "newdefault", "live.params_in1", 5, 117, "live.params_in")
  }

function getcoords(a){
	var f = new File(a);
	var i,r,e, coords;

	if (f.isopen) {
		i=0;
		while (((a = f.readline()) != null) && i<200) { // returns a string
			r = a.search('\"rect\"');
			if(r > -1){
				e = a.lastIndexOf("]");
				a = a.slice(r+11,e-1);
				coords = a.split(',');
				break;
			}
			i++;
		}
		f.close();
	} else {
		post("could not open file: " + s + "\n");
	}
	//post("get",coords, "\n");
	return coords;
}
}

**Comment by klausfilip** [comment]
@the-steech 
its in branch new-createact_js
actmaker, and the js's are updated
please test

**Comment by the-steech** [comment]
@klausfilip really nice work, love all the changes!!!  much simpler :)  using the .maxpat file for the act's size works brilliantly, we can just do away with all the extra "act_size.json" nonsense

i went ahead and commented out the "live_extras" function.  with the control@ changes, we can just remove those extra acts "live.midi_in" and "live.param_in" acts :).   i've also remove these files from the repo

additionally, i've created an "info" window with basic rundown of how to use

is it ok if i commit these changes directly to new-createact_js branch?


<img width="501" alt="image" src="https://user-images.githubusercontent.com/114934553/220983496-90398234-4aad-4f68-9c38-c90a38ad6d6f.png">

**Comment by klausfilip** [comment]
"is it ok if i commit these changes directly to new-createact_js branch?" 
oh yes, please ;)

**Comment by klausfilip** [comment]
actually there is one issue (at least)
that tetris is not loaded with the environment in live.
i'll have a look.

**Comment by the-steech** [comment]
@klausfilip you should make your changes for tetris, then i'll add my changes and we can merge to main :)

**Comment by klausfilip** [comment]
done ;)

**Comment by the-steech** [comment]
@klausfilip I'm also going to remove "act_sizes.json" and related code in tetris@, as well as "ll.live.initialize_bpatcher.js"

**Comment by klausfilip** [comment]
ok, there was this function name() in tetrishelp.js, that stumbled me..

**Reply by the-steech** [reply]
it doesn't seem to be in used by anything, at least there is no "name" on the route object above it.  going to comment out and test

**Comment by klausfilip** [comment]
ja, but it can be called by "getpatcher name"

**Reply by the-steech** [reply]
i can't find that message in actmaker.  do you think we should leave it in?

**Comment by klausfilip** [comment]
obviously it was one of your act_sizes.json hacks, 
so delete it.

**Comment by the-steech** [comment]
all done, should be ready to merge

**Comment by klausfilip** [comment]
hmm, i did the merge, but something went wrong (again).
main is now behind our branch, but does claim to be the same..
(manually comparing tetrishelp.js)

**Reply by the-steech** [reply]
it looks ok to me, it looks like the merge was fine and main is the last updated (on my end)

**Comment by klausfilip** [comment]
if i open tetrishelp.js from main 
there is still the function name(name)
maybe its my Github desktop...

**Reply by the-steech** [reply]
that's weird, it's definitely out of both main and the new-createact_js branch (on my computer) 🤷

**Comment by klausfilip** [comment]
all good,
recloned and its fine now.
max was opening files from trash
...i will take a break from ppooll_dev now ;)
would like to close this discussion, but no option..,
should we delete it?

**Comment by the-steech** [comment]
i'm going to lock it, i think we should keep it around while we are still sorta "testing" live.ppooll, it might be helpful as a reference in the future.

thanks for all this, klaus!!!

**Comment by klausfilip** [comment]
thanks to you my dear !!
